### PR TITLE
Make the window name `_quake` special

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

This PR adds some special behavior to the window named "\_quake".
* When creating the quake window, it ignores "initialRows" and "initialCols" and opens on the top half of the monitor.
  - It uses `initialPosition` to determine which monitor this is
* It cannot be moved
* It can only be vertically resized on the bottom border.
* It's always in focus mode.
  - We should probably have an issue tracking "Allow showing tabs in focus mode"? Maybe?
  - This one element is maybe the one I'm least attached to

When renaming a window to "\_quake", it adopts all those behaviors as well. It does not exit focus mode when leaving QM, nor does it resize back. That seemed unnecessary. 

## References

* As spec'ed in #9274
* See also #8888

## PR Checklist
* [x] In the pursuit of #653 
* [x] I work here
* [ ] Tests added/passed
* [ ] Requires documentation to be updated, but I'm not gonna do any of that till quake mode is totally done. 

## Detailed Description of the Pull Request / Additional comments

Note that this doesn't do things like:
* dropdown
* global hotkey summon 
* summon to the current monitor 
* summon to the current desktop

I'm doing #653 _very_ piecemeal, to try and make the PRs less egregious.

## Validation Steps Performed

* validated that center on launch still works
* validated that QM works on different monitors based on `initialPosition`
* validated entering/exiting QM behaves as expected

## TODO!
* [ ] When snapping the quake window between desktops with <kbd>win+shift+arrow</kbd>, the window doesn't horizontally re-size to the new monitor dimensions. It should.

### CODE REVIEW

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 1]
```diff
@@ -2735,4 +2754,9 @@ namespace winrt::TerminalApp::implementation
             WindowRenamer().IsOpen(false);
         }
     }
+
+    bool TerminalPage::IsQuakeWindow() const
+    {
+        return WindowName() == L"_quake";
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2760]: nit: maybe move `L"_quake" to the top as a `static constexpr std::wstring_view`?

File: src/cascadia/TerminalApp/AppLogic.cpp [hunk 2]
```diff
@@ -286,11 +286,16 @@ namespace winrt::TerminalApp::implementation
             // that the window size is _first_ set up as something sensible, so
             // leaving fullscreen returns to a reasonable size.
             const auto launchMode = this->GetLaunchMode();
-            if (launchMode == LaunchMode::FullscreenMode)
+            if (IsQuakeWindow())
+            {
+                _root->ToggleFocusMode();
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/AppLogic.cpp:291]: Why `ToggleFocusMode` and not something like `EnableFocusMode` or `SetFocusMode(true)`? QM will always be in focus mode right?

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 3]
```diff
@@ -195,6 +195,16 @@ LRESULT IslandWindow::_OnSizing(const WPARAM wParam, const LPARAM lParam)
 
     auto clientHeight = winRect->bottom - winRect->top - nonClientSize.cy;
 
+    // If we're the quake window, prevent resizing on all sides except the
+    // bottom. This also applies to resising with the Alt+Space menu
```
Comment (carlos-zamora) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:199]:
```suggestion
    // bottom. This also applies to resizing with the Alt+Space menu
```

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 4]
```diff
@@ -195,6 +195,16 @@ LRESULT IslandWindow::_OnSizing(const WPARAM wParam, const LPARAM lParam)
...
     auto clientHeight = winRect->bottom - winRect->top - nonClientSize.cy;
 
+    // If we're the quake window, prevent resizing on all sides except the
+    // bottom. This also applies to resising with the Alt+Space menu
+    if (IsQuakeWindow() && wParam != WMSZ_BOTTOM)
+    {
+        // Stuff our current window size into the lParam, and return true. This
+        // will tell User32 to use our current dimensions to resize to.
+        ::GetWindowRect(_window.get(), winRect);
+        return true;
```
Comment (carlos-zamora) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:205]:
```suggestion
        return TRUE;
```
nit: probably doesn't matter, but you return `TRUE` on line 257 so figured you'd want to be consistent

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 5]
```diff
@@ -880,5 +919,63 @@ winrt::fire_and_forget IslandWindow::SummonWindow()
...
+void IslandWindow::_enterQuakeMode()
+{
+    RECT windowRect = GetWindowRect();
+    HMONITOR hmon = MonitorFromRect(&windowRect, MONITOR_DEFAULTTONEAREST);
+    MONITORINFO nearestMonitorInfo;
+
+    UINT dpix = USER_DEFAULT_SCREEN_DPI;
+    UINT dpiy = USER_DEFAULT_SCREEN_DPI;
+    // If this fails, we'll use the default of 96.
+    GetDpiForMonitor(hmon, MDT_EFFECTIVE_DPI, &dpix, &dpiy);
```
Comment (carlos-zamora) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:948]: Sounds like you need to check the returned HRESULT here haha

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 6]
```diff
@@ -880,5 +919,63 @@ winrt::fire_and_forget IslandWindow::SummonWindow()
...
+
+    nearestMonitorInfo.cbSize = sizeof(MONITORINFO);
+    // Get monitor dimensions:
+    GetMonitorInfo(hmon, &nearestMonitorInfo);
```
Comment (carlos-zamora) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:952]: Should we check the return value for this function?

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 7]
```diff
@@ -880,5 +919,63 @@ winrt::fire_and_forget IslandWindow::SummonWindow()
...
+    const til::size desktopDimensions{ gsl::narrow<short>(nearestMonitorInfo.rcWork.right - nearestMonitorInfo.rcWork.left),
+                                       gsl::narrow<short>(nearestMonitorInfo.rcWork.bottom - nearestMonitorInfo.rcWork.top) };
```
Comment (carlos-zamora) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:954]: nit: I think we're trying to use chromium math instead of gsl now.

Comment (carlos-zamora) [reply src/cascadia/WindowsTerminal/IslandWindow.cpp:954]: Like, we can use `base::ClampedNumeric<short>` or `base::CheckedNumeric<short>` depending on what you want.

File: src/cascadia/TerminalApp/TabManagement.cpp [hunk 8]
```diff
@@ -272,12 +272,17 @@ namespace winrt::TerminalApp::implementation
...
+        {
+            // collapse/show the tabs themselves
+            _tabView.Visibility(isVisible ? Visibility::Visible : Visibility::Collapsed);
+        }
+        if (_tabRow)
+        {
+            // collapse/show the row that the tabs are in.
+            // NaN is the special value XAML uses for "Auto" sizing.
+            _tabRow.Height(isVisible ? NAN : 0);
+        }
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TabManagement.cpp:285]: This seems a bit unrelated. Did QM initialize without a `_tabView` and `_tabRow` or what happened?

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 9]
```diff
@@ -2260,7 +2268,7 @@ namespace winrt::TerminalApp::implementation
 
     bool TerminalPage::FocusMode() const
     {
-        return _isInFocusMode;
+        return _isInFocusMode || IsQuakeWindow();
     }
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2272]: This seems a bit weird to me. If QM is always in FM, why do we need to make this change at all? Shouldn't `_isInFocusMode` be updated to be `true` when in QM, and `false` when out of it?

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 10]
```diff
@@ -2596,10 +2604,21 @@ namespace winrt::TerminalApp::implementation
...
             _WindowName = value;
             _PropertyChangedHandlers(*this, WUX::Data::PropertyChangedEventArgs{ L"WindowNameForDisplay" });
+
+            // If we're entering quake mode, or leaving it
+            if (IsQuakeWindow() != oldIsQuakeMode)
+            {
+                // If we're entering QM from ~FM, then this will enter FM
+                // If we're entering QM from FM, then this will do nothing
+                // If we're leaving QM (we're already in FM), then this will do nothing
+                _SetFocusMode(_isInFocusMode);
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2619]: A continuation of my confusion from above, why not do `ToggleFocusMode`?

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 11]
```diff
@@ -242,7 +244,6 @@ void AppHost::Initialize()
     _logic.FullscreenChanged({ this, &AppHost::_FullscreenChanged });
     _logic.FocusModeChanged({ this, &AppHost::_FocusModeChanged });
     _logic.AlwaysOnTopChanged({ this, &AppHost::_AlwaysOnTopChanged });
-    _logic.RaiseVisualBell({ this, &AppHost::_RaiseVisualBell });
```
Comment (carlos-zamora) [inline src/cascadia/WindowsTerminal/AppHost.cpp:245]: I think we still want to continue raising the visual bell right? haha

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 12]
```diff
@@ -57,6 +57,8 @@ AppHost::AppHost() noexcept :
         _window = std::make_unique<IslandWindow>();
     }
 
+    _window->IsQuakeWindow(_logic.IsQuakeWindow());
```
Comment (carlos-zamora) [inline src/cascadia/WindowsTerminal/AppHost.cpp:60]: Should we be calling `_IsQuakeWindowChanged` instead?

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 13]
```diff
@@ -379,39 +381,59 @@ void AppHost::_HandleCreateWindow(const HWND hwnd, RECT proposedRect, LaunchMode
...
+            availableSpace.height() / 2
+        };
+        launchMode = LaunchMode::FocusMode;
     }
-    const COORD origin{ gsl::narrow<short>(proposedRect.left),
-                        gsl::narrow<short>(proposedRect.top) };
 
-    const auto newPos = Viewport::FromDimensions(origin, dimensions);
+    const til::rectangle newRect{ origin, dimensions };
+    // const auto newPos = Viewport::FromDimensions(origin, dimensions);
```
Comment (carlos-zamora) [inline src/cascadia/WindowsTerminal/AppHost.cpp:430]: dead code

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 14]
```diff
@@ -880,5 +919,63 @@ winrt::fire_and_forget IslandWindow::SummonWindow()
...
+    {
+        _isQuakeWindow = isQuakeWindow;
+        if (IsQuakeWindow())
+        {
+            _enterQuakeMode();
+        }
+    }
+}
+
+void IslandWindow::_enterQuakeMode()
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:939]: I don't see any monarch changes here... is something else stopping two windows from getting named `_quake`? You did say that someone could end up naming it that themselves and it would get the behavior reserved here... but I would think that ending up with multiple of them is bad. Also no idea how you'll block that cross-elevation...

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 15]
```diff
@@ -379,39 +381,59 @@ void AppHost::_HandleCreateWindow(const HWND hwnd, RECT proposedRect, LaunchMode
...
-                            ((desktopDimensions.X / 2) - (dimensions.X / 2));
-        proposedRect.top = nearestMonitorInfo.rcWork.top +
-                           ((desktopDimensions.Y / 2) - (dimensions.Y / 2));
+        origin = til::point{
+            nearestMonitorInfo.rcWork.left + ((desktopDimensions.width() / 2) - (dimensions.width() / 2)),
+            nearestMonitorInfo.rcWork.top + ((desktopDimensions.height() / 2) - (dimensions.height() / 2))
+        };
+    }
+
+    if (_logic.IsQuakeWindow())
```
Comment (carlos-zamora) [inline src/cascadia/WindowsTerminal/AppHost.cpp:411]: nit: should we check for `IsQuakeWindow` first, then `else if (centerOnLaunch)`? It's minor but I guess we could save a bit of time with having to calculate the origin again.

**Review by carlos-zamora — CHANGES_REQUESTED** [review]
A lot of nits, but a few questions here and there that I want answers on before signing off.

File: src/cascadia/WindowsTerminal/IslandWindow.cpp (hunk 14 above)
Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/IslandWindow.cpp:939]: 1. The window naming PR already prevented windows from having the same name
2. there's nothing that we can do to prevent having simultaneously an elevated and unelevated window with the same name. So you can have an Admin: \_quake and an unelevated one. It's a shame, but packaged COM & mixed elevation really don't play well 😢

File: src/cascadia/WindowsTerminal/AppHost.cpp (hunk 12 above)
Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/AppHost.cpp:60]: Nah, because `_IsQuakeWindowChanged` assumes that the window actually already exists. This is before the HWND has actually been created. `_HandleCreateWindow` will do the right thing for quake mode when the window is first created.

Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/AppHost.cpp:60]: Ignore me, I was thinking of `_enterQuakeMode`. Yea sure we can do _IsQuakeWindowChanged

File: src/cascadia/WindowsTerminal/AppHost.cpp (hunk 11 above)
Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/AppHost.cpp:245]: must've been a bad merge ¯\\\_(ツ)_/¯

File: src/cascadia/WindowsTerminal/IslandWindow.cpp (hunk 4 above)
Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/IslandWindow.cpp:205]: I'm actually gonna swing the other way and make them lower-case

File: src/cascadia/TerminalApp/TabManagement.cpp (hunk 8 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TabManagement.cpp:285]: > Did QM initialize without a `_tabView` and `_tabRow`

SURE DID

File: src/cascadia/TerminalApp/AppLogic.cpp (hunk 2 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/AppLogic.cpp:291]: Frankly, I hadn't written `_SetFocusMode` yet when I added this line 😋

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/AppLogic.cpp:291]: Ah, but that's a private member on TerminalPage... meh, I just didn't feel like adding another method that'll only have one usage right here

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 9 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2272]: yea that's weird

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 10 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2619]: meh, this one gets to be more explicit in what it's trying to do because it's already inside the TerminalPage. Technically, yes, `ToggleFocusMode` would work, by setting it to false, but then staying in focus mode anyways, but ¯\\\_(ツ)_/¯

**Review by carlos-zamora — APPROVED** [review]
Looks good to me! Thanks!

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 16]
```diff
@@ -416,18 +410,25 @@ void AppHost::_HandleCreateWindow(const HWND hwnd, RECT proposedRect, LaunchMode
         const til::size availableSpace = desktopDimensions + nonClientSize;
 
         origin = til::point{
-            nearestMonitorInfo.rcWork.left - (nonClientSize.width() / 2),
-            nearestMonitorInfo.rcWork.top
+            ::base::saturated_cast<ptrdiff_t>(nearestMonitorInfo.rcWork.left - (nonClientSize.width() / 2)),
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/AppHost.cpp:413]: I feel like this is weird. We're doing the math unprotected from over/underflow first then casting the result into `ptrdiff_t`.

Shouldn't we be doing a saturated operation on these things or casting them into the math wrappers first and then let it do the saturated subtract? 

The same goes for some of the other saturated casts below. Though I'm not totally clear on what the source types are here that requires us to do the cast to begin with.

Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/AppHost.cpp:413]: Yea you know, I felt weird writing that. I thinkthe types are LONGs originally. Maybe `point` is more forgiving than `size`, which is why I was getting weird "can't use an initializer list" errors. 

lemme back that out and spend more than 30 seconds on it

**Review by miniksa — APPROVED** [review]
That's an excellent idea for fixing the math thing. Appreciated. Looks good to me in that respect and otherwise.
